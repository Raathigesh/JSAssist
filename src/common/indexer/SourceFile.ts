import {
  ExportDefaultDeclaration,
  ExportNamedDeclaration,
  ImportDeclaration
} from "babel-types";
import { readFile } from "fs";
import { promisify } from "util";
import * as parser from "@babel/parser";
import traverse from "@babel/traverse";
import { NodePath } from "babel-traverse";
import FunctionDefinition from "./FunctionDefinition";
import ClassDefinition from "./classDefinition";
import { getFileType } from "../../extension/utils/file";
import VariableDeclaration from "./VariableDeclaration";
import ImportStatement from "./ImportStatement";

export default class SourceFile {
  public path: string = "";
  public functions: FunctionDefinition[] = [];
  public classes: ClassDefinition[] = [];
  public variables: VariableDeclaration[] = [];
  public importStatements: ImportStatement[] = [];

  public async parse(filePath: string) {
    this.path = filePath;

    const content = await promisify(readFile)(filePath);
    const ast = this.getAST(content.toString(), filePath);

    traverse(ast, {
      ExportNamedDeclaration: (path: NodePath<ExportNamedDeclaration>) => {
        this.extractNamedExport(path);
      },
      ExportDefaultDeclaration: (path: NodePath<ExportDefaultDeclaration>) => {
        this.extractDefaultExport(path);
      },
      ImportDeclaration: (path: NodePath<ImportDeclaration>) => {
        this.extractImportStatements(path);
      }
    });
  }

  public isImportingSymbol(name: string, source: string) {
    return this.importStatements
      .filter(item => item.path === source)
      .some(statement => {
        return (
          statement.defaultImportName == name ||
          statement.namedImports.some(
            namedImport =>
              namedImport.name === name || namedImport.aliasName === name
          )
        );
      });
  }

  private extractImportStatements(path: NodePath<ImportDeclaration>) {
    const resolveMap = {
      "app-simple-plans": "./src/app-simple-plans"
    };

    const source = path.node.source.value;
    const importStatement = new ImportStatement();

    let fullSourcePath = source;
    for (const [alias, relativePath] of Object.entries(resolveMap)) {
      if (fullSourcePath.startsWith(alias)) {
        fullSourcePath = fullSourcePath.replace(alias, relativePath);
      }
    }
    importStatement.path = fullSourcePath;

    path.node.specifiers.map(specifier => {
      if (specifier.type === "ImportDefaultSpecifier") {
        importStatement.defaultImportName = specifier.local.name;
      } else if (specifier.type === "ImportSpecifier") {
        importStatement.namedImports.push({
          name: specifier.local.name
        });
      }
    });

    this.importStatements.push(importStatement);
  }

  private extractDefaultExport(path: NodePath<ExportDefaultDeclaration>) {
    const declaration = path.node.declaration;
    if (declaration && declaration.type === "FunctionDeclaration") {
      const functionDefinition = new FunctionDefinition();
      functionDefinition.name = (declaration.id && declaration.id.name) || "";
      functionDefinition.exportStatus = "default";
      functionDefinition.location = path.node.loc;
      this.functions.push(functionDefinition);
    } else if (declaration && declaration.type === "ClassDeclaration") {
      const classDefinition = new ClassDefinition();
      classDefinition.name = (declaration.id && declaration.id.name) || "";
      classDefinition.exportStatus = "default";
      classDefinition.location = path.node.loc;
      this.classes.push(classDefinition);
    } else if (declaration && declaration.type === "VariableDeclaration") {
      const variableDeclaration = new VariableDeclaration();
      variableDeclaration.name =
        (declaration.declarations[0] &&
          declaration.declarations[0].id &&
          (declaration.declarations[0].id as any).name) ||
        "";
      variableDeclaration.exportStatus = "default";
      variableDeclaration.location = path.node.loc;
      this.variables.push(variableDeclaration);
    }
  }

  private extractNamedExport(path: NodePath<ExportNamedDeclaration>) {
    const declaration = path.node.declaration;
    if (declaration && declaration.type === "FunctionDeclaration") {
      const functionDefinition = new FunctionDefinition();
      functionDefinition.name = (declaration.id && declaration.id.name) || "";
      functionDefinition.exportStatus = "named";
      functionDefinition.location = path.node.loc;
      this.functions.push(functionDefinition);
    } else if (declaration && declaration.type === "ClassDeclaration") {
      const classDefinition = new ClassDefinition();
      classDefinition.name = (declaration.id && declaration.id.name) || "";
      classDefinition.exportStatus = "named";
      classDefinition.location = path.node.loc;
      this.classes.push(classDefinition);
    } else if (declaration && declaration.type === "VariableDeclaration") {
      const variableDeclaration = new VariableDeclaration();
      variableDeclaration.name =
        (declaration.declarations[0] &&
          declaration.declarations[0].id &&
          (declaration.declarations[0].id as any).name) ||
        "";
      variableDeclaration.exportStatus = "named";
      variableDeclaration.location = path.node.loc;
      this.variables.push(variableDeclaration);
    }
  }

  private getAST(content: string, path: string) {
    const fileType = getFileType(path);

    const isTS = fileType === "TSX";
    const additionalPlugin = isTS ? "typescript" : "flow";

    return parser.parse(content, {
      sourceType: "module",
      plugins: ["jsx", "classProperties", "dynamicImport", additionalPlugin]
    });
  }
}
