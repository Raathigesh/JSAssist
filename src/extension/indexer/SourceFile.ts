import { ExportDefaultDeclaration, ExportNamedDeclaration } from "babel-types";
import { readFile } from "fs";
import { promisify } from "util";
import * as parser from "@babel/parser";
import traverse from "@babel/traverse";
import { NodePath } from "babel-traverse";
import FunctionDefinition from "./FunctionDefinition";
import ClassDefinition from "./classDefinition";
import { getFileType } from "../utils/file";
import VariableDeclaration from "./VariableDeclaration";

export default class SourceFile {
  public functions: FunctionDefinition[] = [];
  public classes: ClassDefinition[] = [];
  public variables: VariableDeclaration[] = [];

  public async parse(filePath: string) {
    const content = await promisify(readFile)(filePath);
    const ast = this.getAST(content.toString(), filePath);

    traverse(ast, {
      ExportNamedDeclaration: (path: NodePath<ExportNamedDeclaration>) => {
        const declaration = path.node.declaration;
        if (declaration && declaration.type === "FunctionDeclaration") {
          const functionDefinition = new FunctionDefinition();
          functionDefinition.name =
            (declaration.id && declaration.id.name) || "";
          functionDefinition.exportStatus = "named";
          this.functions.push(functionDefinition);
        } else if (declaration && declaration.type === "ClassDeclaration") {
          const classDefinition = new ClassDefinition();
          classDefinition.name = (declaration.id && declaration.id.name) || "";
          classDefinition.exportStatus = "named";
          this.classes.push(classDefinition);
        } else if (declaration && declaration.type === "VariableDeclaration") {
          const variableDeclaration = new VariableDeclaration();
          variableDeclaration.name =
            (declaration.declarations[0] &&
              declaration.declarations[0].id &&
              (declaration.declarations[0].id as any).name) ||
            "";
          variableDeclaration.exportStatus = "named";
          this.variables.push(variableDeclaration);
        }
      },
      ExportDefaultDeclaration: (path: NodePath<ExportDefaultDeclaration>) => {
        const declaration = path.node.declaration;
        if (declaration && declaration.type === "FunctionDeclaration") {
          const functionDefinition = new FunctionDefinition();
          functionDefinition.name =
            (declaration.id && declaration.id.name) || "";
          functionDefinition.exportStatus = "default";
          this.functions.push(functionDefinition);
        } else if (declaration && declaration.type === "ClassDeclaration") {
          const classDefinition = new ClassDefinition();
          classDefinition.name = (declaration.id && declaration.id.name) || "";
          classDefinition.exportStatus = "default";
          this.classes.push(classDefinition);
        } else if (declaration && declaration.type === "VariableDeclaration") {
          const variableDeclaration = new VariableDeclaration();
          variableDeclaration.name =
            (declaration.declarations[0] &&
              declaration.declarations[0].id &&
              (declaration.declarations[0].id as any).name) ||
            "";
          variableDeclaration.exportStatus = "default";
          this.variables.push(variableDeclaration);
        }
      }
    });
  }

  private getAST(content: string, path: string) {
    const fileType = getFileType(path);

    const isTS = fileType === "TSX";
    const additionalPlugin = isTS ? "typescript" : "flow";

    return parser.parse(content, {
      sourceType: "module",
      plugins: ["jsx", "classProperties", "dynamicImport", additionalPlugin]
    });
  }
}
