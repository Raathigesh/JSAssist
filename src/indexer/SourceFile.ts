import {
  FunctionDeclaration,
  VariableDeclaration,
  ClassDeclaration,
  TypeAlias,
  SourceLocation,
  Identifier,
  Program,
  ExportNamedDeclaration
} from "babel-types";
import { readFile } from "fs";
import { promisify } from "util";
const nanoid = require("nanoid");
import * as parser from "@babel/parser";
import traverse from "@babel/traverse";
import { NodePath, Scope } from "babel-traverse";
import { getFileType } from "indexer/util";
import ImportStatement, { ImportSpecifier } from "./ImportStatement";
import ESModuleItem from "./ESModuleItem";
import { ImportDeclaration } from "@babel/types";
import { dirname } from "path";
import { findAbsoluteFilePathWhichExists } from "./fileResolver";
import { GqlLocation } from "entities/GqlLocation";
import ExportStatement from "./ExportStatement";

export default class SourceFile {
  public path: string = "";
  public symbols: ESModuleItem[] = [];
  public importStatements: ImportStatement[] = [];
  public exportStatements: ExportStatement[] = [];
  public programScope: Scope | undefined;

  public async parse(
    filePath: string,
    pathAliasMap: { [alias: string]: string },
    root: string
  ) {
    try {
      this.path = filePath;

      const content = await promisify(readFile)(filePath);
      const ast = this.getAST(content.toString(), filePath);

      traverse(ast, {
        Program: (path: NodePath<Program>) => {
          this.programScope = path.scope;
        },
        FunctionDeclaration: (path: NodePath<FunctionDeclaration>) => {
          this.extractFunctionDeclaration(path);
        },
        VariableDeclaration: (path: NodePath<VariableDeclaration>) => {
          this.extractVariableDeclaration(path);
        },
        ClassDeclaration: (path: NodePath<ClassDeclaration>) => {
          this.extractClassDeclaration(path);
        },
        TypeAlias: (path: NodePath<TypeAlias>) => {
          this.extractTypeAlias(path);
        },
        ImportDeclaration: (path: NodePath<ImportDeclaration>) => {
          this.extractImport(path, pathAliasMap, root);
        },
        ExportNamedDeclaration: (path: NodePath<ExportNamedDeclaration>) => {
          this.extractExportNamedDeclaration(path);
        }
      });
      this.linkLocalSymbols();
    } catch (e) {
      console.log("Parsing failed", filePath, e);
    }
  }

  public getSymbolInPosition(position: GqlLocation) {
    return this.symbols.find(
      sym => sym.location && this.isInLocation(sym.location, position)
    );
  }

  private extractFunctionDeclaration(path: NodePath<FunctionDeclaration>) {
    const name = path.node.id.name;
    this.createSymbol(name, path.node.type, path.node.loc);
  }

  private extractVariableDeclaration(path: NodePath<VariableDeclaration>) {
    if (
      path.parent.type !== "Program" &&
      path.parent.type !== "ExportNamedDeclaration"
    ) {
      return;
    }
    path.getStatementParent;
    const name = (path.node.declarations[0].id as Identifier).name;
    this.createSymbol(name, path.node.type, path.node.loc);
  }

  private extractClassDeclaration(path: NodePath<ClassDeclaration>) {
    const name = path.node.id.name;
    this.createSymbol(name, path.node.type, path.node.loc);
  }

  private extractTypeAlias(path: NodePath<TypeAlias>) {
    const name = path.node.id.name;
    this.createSymbol(name, path.node.type, path.node.loc);
  }

  private extractExportNamedDeclaration(
    path: NodePath<ExportNamedDeclaration>
  ) {
    if (path.node.source && path.node.specifiers.length) {
      const exportStatement: ExportStatement = {
        path: path.node.source.value,
        specifiers: path.node.specifiers.map(specifier => ({
          local: specifier.local.name,
          exported: specifier.exported.name
        }))
      };
      this.exportStatements.push(exportStatement);
    }
  }

  private createSymbol(name: string, kind: string, location: SourceLocation) {
    const symbol = new ESModuleItem();
    symbol.id = nanoid();
    symbol.name = name;
    symbol.kind = kind;
    symbol.location = location;
    symbol.path = this.path;
    this.importStatements.forEach(importStatement => {
      importStatement.specifiers.forEach(specifier => {
        specifier.references.forEach(reference => {
          if (
            reference.location &&
            this.isInLocation(location, reference.location)
          ) {
            symbol.markers.push({
              filePath: importStatement.path,
              name: specifier.name,
              location: this.adjustLocation(location, reference.location) as any
            });
          }
        });
      });
    });
    this.symbols.push(symbol);
  }

  private linkLocalSymbols() {
    if (!this.programScope) {
      return;
    }

    const bindings = Object.entries(this.programScope.bindings);
    bindings.forEach(([name, binding]) => {
      binding.referencePaths.forEach(reference => {
        this.symbols.forEach(symbol => {
          if (
            reference.node.loc &&
            symbol.location &&
            this.isInLocation(symbol.location, reference.node.loc)
          ) {
            const existingMarker = symbol.markers.find(
              marker => marker.name === name
            );

            if (!existingMarker) {
              symbol.markers.push({
                filePath: symbol.path,
                name,
                location: this.adjustLocation(
                  symbol.location,
                  reference.node.loc
                ) as any
              });
            }
          }
        });
      });
    });
  }

  private extractImport(
    path: NodePath<ImportDeclaration>,
    pathAliasMap: { [alias: string]: string },
    root: string
  ) {
    const importPath = path.node.source.value;
    let absoluteImportPath = importPath;
    try {
      absoluteImportPath = findAbsoluteFilePathWhichExists(
        root,
        dirname(this.path),
        importPath,
        pathAliasMap
      );
      console.log("found path ", absoluteImportPath);
    } catch (e) {
      console.log("Path not resolved ", importPath);
      return;
    }

    const importDeclaration: ImportStatement = {
      path: absoluteImportPath,
      specifiers: []
    };

    path.node.specifiers.forEach(specifierPath => {
      const referencePaths =
        path.scope.bindings[specifierPath.local.name].referencePaths;
      const defaultSpecifier: ImportSpecifier = {
        name:
          ((specifierPath as any).imported &&
            (specifierPath as any).imported.name) ||
          specifierPath.local.name,
        references: referencePaths.map(referencePath => {
          return {
            location: referencePath.node.loc
          };
        })
      };
      importDeclaration.specifiers.push(defaultSpecifier);
    });
    this.importStatements.push(importDeclaration);
  }

  private getAST(content: string, path: string) {
    const fileType = getFileType(path);

    const isTS = fileType === "TSX";
    const additionalPlugin = isTS ? "typescript" : "flow";

    return parser.parse(content, {
      sourceType: "module",
      plugins: ["jsx", "classProperties", "dynamicImport", additionalPlugin]
    });
  }

  private isInLocation(locationA: GqlLocation, locationB: GqlLocation) {
    const {
      start: locAStart = { line: 0, column: 0 },
      end: locAEnd = { line: 0, column: 0 }
    } = locationA;

    const {
      start: locBStart = { line: 0, column: 0 },
      end: locBEnd = { line: 0, column: 0 }
    } = locationB;

    let isStartWithIn = false;
    if (locBStart.line > locAStart.line) {
      isStartWithIn = true;
    } else if (locBStart.line === locAStart.line) {
      isStartWithIn = locBStart.column >= locAStart.column;
    }

    let isEndWithIn = false;
    if (locBEnd.line < locAEnd.line) {
      isEndWithIn = true;
    } else if (locBEnd.line === locAEnd.line) {
      isEndWithIn = locBEnd.column <= locAEnd.column;
    }

    return isStartWithIn && isEndWithIn;
  }

  private adjustLocation(
    functionLocation: GqlLocation,
    markerLocation: GqlLocation
  ) {
    const {
      start: functionStart = { line: 0, column: 0 },
      end: functionEnd = { line: 0, column: 0 }
    } = functionLocation;

    const {
      start: markerStart = { line: 0, column: 0 },
      end: markerEnd = { line: 0, column: 0 }
    } = markerLocation;

    const newLocation: GqlLocation = {
      start: {
        line: markerStart.line - functionStart.line,
        column: markerStart.column
      },
      end: {
        line: markerEnd.line - functionStart.line,
        column: markerEnd.column
      }
    };

    return newLocation;
  }
}
