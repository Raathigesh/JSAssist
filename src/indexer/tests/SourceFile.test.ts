import SourceFile from "../SourceFile";
import { resolve } from "path";

describe("SourceFile", () => {
  it("should have named exports", async () => {
    const sourceFile = new SourceFile();
    await sourceFile.parse(resolve(__dirname, "./javascript.js"));

    expect(sourceFile.symbols).toMatchInlineSnapshot(`
      Array [
        ESModuleItem {
          "exportStatus": "named",
          "id": "",
          "location": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 5,
            },
            "start": Position {
              "column": 0,
              "line": 3,
            },
          },
          "name": "Hello",
          "type": "FunctionDeclaration",
        },
        ESModuleItem {
          "exportStatus": "named",
          "id": "",
          "location": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 9,
            },
            "start": Position {
              "column": 0,
              "line": 7,
            },
          },
          "name": "AnotherFunction",
          "type": "FunctionDeclaration",
        },
        ESModuleItem {
          "exportStatus": "named",
          "id": "",
          "location": SourceLocation {
            "end": Position {
              "column": 2,
              "line": 13,
            },
            "start": Position {
              "column": 0,
              "line": 11,
            },
          },
          "name": "AnotherUtil",
          "type": "VariableDeclaration",
        },
      ]
    `);

    expect(sourceFile.importStatements).toMatchInlineSnapshot(`
      Array [
        Object {
          "path": "./date",
          "specifiers": Array [
            Object {
              "isDefault": true,
              "name": "getDate",
              "references": Array [
                Object {
                  "containerName": "Hello",
                  "containerType": "FunctionDeclaration",
                },
                Object {
                  "containerName": "AnotherFunction",
                  "containerType": "FunctionDeclaration",
                },
                Object {
                  "containerName": "AnotherUtil",
                  "containerType": "FunctionDeclaration",
                },
              ],
            },
          ],
        },
      ]
    `);
  });
});
