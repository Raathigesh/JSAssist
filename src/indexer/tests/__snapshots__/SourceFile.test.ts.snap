// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceFile should have named exports 1`] = `
Array [
  Object {
    "id": "",
    "kind": "FunctionDeclaration",
    "location": SourceLocation {
      "end": Position {
        "column": 29,
        "line": 1,
      },
      "start": Position {
        "column": 7,
        "line": 1,
      },
    },
    "markers": Array [],
    "name": "getItems",
    "path": "root/query/items.js",
  },
  Object {
    "id": "",
    "kind": "FunctionDeclaration",
    "location": SourceLocation {
      "end": Position {
        "column": 36,
        "line": 2,
      },
      "start": Position {
        "column": 7,
        "line": 2,
      },
    },
    "markers": Array [],
    "name": "getInStockCount",
    "path": "root/query/items.js",
  },
]
`;

exports[`SourceFile should have named exports 2`] = `Array []`;

exports[`SourceFile should identify markers 1`] = `
Array [
  Object {
    "id": "",
    "kind": "VariableDeclaration",
    "location": SourceLocation {
      "end": Position {
        "column": 29,
        "line": 4,
      },
      "start": Position {
        "column": 0,
        "line": 4,
      },
    },
    "markers": Array [],
    "name": "getSessions",
    "path": "root/store/Root.js",
  },
  Object {
    "id": "",
    "kind": "FunctionDeclaration",
    "location": SourceLocation {
      "end": Position {
        "column": 1,
        "line": 10,
      },
      "start": Position {
        "column": 15,
        "line": 6,
      },
    },
    "markers": Array [
      Object {
        "filePath": "root/query/items.js",
        "location": Object {
          "end": Object {
            "column": 24,
            "line": 1,
          },
          "start": Object {
            "column": 16,
            "line": 1,
          },
        },
        "name": "getItems",
      },
      Object {
        "filePath": "project/query/items",
        "location": Object {
          "end": Object {
            "column": 36,
            "line": 2,
          },
          "start": Object {
            "column": 21,
            "line": 2,
          },
        },
        "name": "getInStockCount",
      },
      Object {
        "filePath": "root/store/Root.js",
        "location": Object {
          "end": Object {
            "column": 30,
            "line": 3,
          },
          "start": Object {
            "column": 19,
            "line": 3,
          },
        },
        "name": "getSessions",
      },
    ],
    "name": "Store",
    "path": "root/store/Root.js",
  },
]
`;

exports[`SourceFile should identify markers 2`] = `
Array [
  Object {
    "path": "root/query/items.js",
    "specifiers": Array [
      Object {
        "isDefault": true,
        "name": "getItems",
        "references": Array [
          Object {
            "code": "",
            "containerName": "Store",
            "containerType": "FunctionDeclaration",
            "location": SourceLocation {
              "end": Position {
                "column": 24,
                "line": 7,
              },
              "identifierName": "getItems",
              "start": Position {
                "column": 16,
                "line": 7,
              },
            },
          },
        ],
      },
    ],
  },
  Object {
    "path": "project/query/items",
    "specifiers": Array [
      Object {
        "isDefault": true,
        "name": "getInStockCount",
        "references": Array [
          Object {
            "code": "",
            "containerName": "Store",
            "containerType": "FunctionDeclaration",
            "location": SourceLocation {
              "end": Position {
                "column": 36,
                "line": 8,
              },
              "identifierName": "getInStockCount",
              "start": Position {
                "column": 21,
                "line": 8,
              },
            },
          },
        ],
      },
    ],
  },
]
`;
